tcmur_cmd_handler.c

```diff
@@ -33,33 +34,42 @@ static void _cleanup_spin_lock(void *arg)
 }
 
 void tcmur_command_complete(struct tcmu_device *dev, struct tcmulib_cmd *cmd,
-			    int rc)
+			    int rc, bool timeout)
 {
 	struct tcmur_device *rdev = tcmu_get_daemon_dev_private(dev);
 
 	pthread_cleanup_push(_cleanup_spin_lock, (void *)&rdev->lock);
-	pthread_spin_lock(&rdev->lock);
-
-	tcmulib_command_complete(dev, cmd, rc);
+	if (!timeout && cmd->timer)
+		tcmu_del_timer(cmd->timer);
 
+	pthread_spin_lock(&rdev->lock);
+	tcmulib_command_complete(dev, cmd, rc, timeout);
 	pthread_spin_unlock(&rdev->lock);
 	pthread_cleanup_pop(0);
 }
 
 static void aio_command_finish(struct tcmu_device *dev, struct tcmulib_cmd *cmd,
-			       int rc)
+			       int rc, bool timeout)
 {
 	struct tcmur_device *rdev = tcmu_get_daemon_dev_private(dev);
 	int wake_up;
 
-	tcmur_command_complete(dev, cmd, rc);
+	tcmur_command_complete(dev, cmd, rc, timeout);
 	track_aio_request_finish(rdev, &wake_up);
 	while (wake_up) {
 		tcmulib_processing_complete(dev);
 		track_aio_wakeup_finish(rdev, &wake_up);
 	}
 }
 
+void tcmur_cmd_timeout(struct tcmu_timer *timer, void *data)
+{
+	struct tcmulib_cmd *cmd = (struct tcmulib_cmd *)data;
+	struct tcmu_device *dev = cmd->dev;
+
+	aio_command_finish(dev, cmd, TCMU_STS_HW_ERR, true);
```

Comment:
I think we will want to use something that is less harsh.

For linux it looks like it is treated as a path error for the normal RW case.

Do you know how windows and ESX handle this error?

Reply:
Oh yeah, ignore this comment. Feel free to keep it as HW_ERR since you are the only user. Maybe you want to change it to LUN_COMM error like is used in the kernel when the tcmu cmd time out fires.